src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,157 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head = null;
+    private Node<T> tail = null;
+    private int size = 0;
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private boolean isInvalidIndex(int index) {
+        return index < 0 || index >= size;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkToTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            linkToTail(value);
+        } else if (!isInvalidIndex(index)) {
+            linkToNode(value, getNodeByIndex(index));
+        }
+        if (isInvalidIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T item: list) {
+            add(item);
+        }
+        return true;
+    }
+
+    private void linkToNode(T elementToInsert, Node<T> successor) {
+        Node<T> predecessor = successor.prev;
+        Node<T> newNode = new Node<>(predecessor, elementToInsert, successor);
+        successor.prev = newNode;
+        if (predecessor == null) {
+            head = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkToTail(T elementToInsert) {
+        Node<T> predecessor = tail;
+        Node<T> newNode = new Node<>(predecessor, elementToInsert, null);
+        tail = newNode;
+        if (predecessor == null) {
+            head = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isInvalidIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
+        }
+        return getNodeByIndex(index).value;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> targetNode;
+        if (index > size / 2) {
+            targetNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                targetNode = targetNode.prev;
+            }
+        } else {
+            targetNode = head;
+            for (int i = 0; i < index; i++) {
+                targetNode = targetNode.next;
+            }
+        }
+        return targetNode;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (isInvalidIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
```

Comment:
since you always throw the same exception it would be better to move it to the method as to reduce code duplication

Reply:
Done